File: tests/e2e/cases/07-enum-constraints/auth_enum/auth_enum_verified.cpp
/*
DEMO: Enum-tag range constraint, automatically derived from the
`enum class AuthResult : uint8_t { Ok = 0, Pending = 1, Failed = 2 }`
declaration.

Expected verdict: VERIFIED.

The function switches over every declared variant and ends with
`__builtin_unreachable()` so any out-of-range discriminant is undefined
behaviour. SAW's crucible-llvm raises an error when the `unreachable`
LLVM instruction is reachable, which means without a precondition
constraining `r` to `{0, 1, 2}` the solver will pick `r = 0xff` and the
case DISPROVES.

`saw-spec-gen gen-verify` is expected to emit

    llvm_precond {{ r <= (2 : [8]) }};

automatically based on the EnumDecl + its three variants. The
underlying-type width (`uint8_t` here) becomes the Cryptol bit width.
*/

#include <cstdint>

enum class AuthResult : uint8_t {
    Ok      = 0,
    Pending = 1,
    Failed  = 2,
};

uint32_t classify(AuthResult r) {
    switch (r) {
        case AuthResult::Ok:      return 0;
        case AuthResult::Pending: return 1;
        case AuthResult::Failed:  return 2;
    }
    __builtin_unreachable();
}

File: tests/e2e/cases/07-enum-constraints/auth_enum/auth_enum_verified.rs
// DEMO: Rust mirror of auth_enum_verified.cpp.
//
// `#[repr(u8)] enum AuthResult { Ok=0, Pending=1, Failed=2 }` lowers
// to an `i8` on the LLVM IR boundary. A safe `match` over all declared
// variants would normally be exhaustive, but to expose the same
// problem the C++ side has — symbolic enum tags out of range — we
// drop into `unreachable_unchecked` on the impossible-at-source-level
// `_` arm.
//
// Expected verdict: VERIFIED, once saw-spec-gen learns to emit an
// enum-range precondition from the mir-json ADT info
// (`adts[*].variants` + `discriminant_bits`). Until then verify-rust.ps1
// generates a bare `x0 <- llvm_fresh_var "x0" (llvm_int 8)` with no
// constraint and SAW returns DISPROVED with `x0 = 3`.

#[repr(u8)]
pub enum AuthResult {
    Ok      = 0,
    Pending = 1,
    Failed  = 2,
}

fn classify(r: AuthResult) -> u32 {
    match r {
        AuthResult::Ok      => 0,
        AuthResult::Pending => 1,
        AuthResult::Failed  => 2,
    }
}
